tests/artemis-test-support/src/main/java/org/apache/activemq/transport/netty/NettyTcpTransport.java

```diff
@@ -222,17 +226,34 @@ public ByteBuf allocateSendBuffer(int size) throws IOException {
       return channel.alloc().ioBuffer(size, size);
    }
 
-   @Override
-   public ChannelFuture send(ByteBuf output) throws IOException {
-      checkConnected();
+   protected final ChannelFuture writeAndFlush(ByteBuf output,
+                                               ChannelPromise promise,
+                                               Function<? super ByteBuf, ? extends ReferenceCounted> bufferTransformer) throws IOException {
+      try {
+         checkConnected();
+      } catch (IOException ioEx) {
+         output.release();
+         throw ioEx;
+      }
       int length = output.readableBytes();
       if (length == 0) {
+         output.release();
          return null;
       }
 
       LOG.trace("Attempted write of: {} bytes", length);
 
-      return channel.writeAndFlush(output);
+      return channel.writeAndFlush(bufferTransformer.apply(output), promise);
+   }
+
+   @Override
+   public ChannelFuture send(ByteBuf output) throws IOException {
+      return writeAndFlush(output, channel.newPromise(), identity());
+   }
+
+   @Override
+   public void sendVoidPromise(ByteBuf output) throws IOException {
+      writeAndFlush(output, channel.voidPromise(), identity());
```

Comment:
Are you doing this because you saw a high number of propmises being created by the default method?

Looking at send (the one that's using channel.newPromise()) this is only used now in stomp, and even there it should be using sendVoidPromise i think.


I'm not sure I understand what's the point of creating a promise if you're not using it. I thought it was the case.

So, shouldn't all the sends be done without a promise (or with a void promise, which is like a static instance / singleton)?

Reply:
Yes we can always use void primise as long as we don't need a callback: if the callback is used void promise won't work

Reply:
I just looked further.. the only place this is being used is on client APIs on testing.

Reply:
nice, so in this case I could just make 1 call without callbacks and no promises. If, in the future, we will need it again we will add a new method

Reply:
lets just leave it.. otherwise you will have to rewrite those client testers